packages/app/src/background/services/injector/InjectorHandler.ts

```diff
@@ -0,0 +1,217 @@
+import { validateMerkleProofSource } from "@src/background/services/validation";
+import ZkIdentityService from "@src/background/services/zkIdentity";
+
+export class InjectorHandler {
+  lockerService: LockerService;
+
+  approvalService: ApprovalService;
+
+  zkProofService: ZkProofService;
+
+  requestManager: RequestManager;
+
+  browserService: BrowserUtils;
+
+  zkIdentityService: ZkIdentityService;
+
+  constructor() {
+    this.approvalService = ApprovalService.getInstance();
+    this.browserService = BrowserUtils.getInstance();
+    this.lockerService = LockerService.getInstance();
+    this.requestManager = RequestManager.getInstance();
+    this.zkIdentityService = ZkIdentityService.getInstance();
+    this.zkProofService = new ZkProofService();
+  }
+
+  connectedIdentityMetadata = async (_: unknown, meta?: IZkMetadata): Promise<ConnectedIdentityMetadata> => {
+    const connectedIdentityMetadata = await this.zkIdentityService.getConnectedIdentityData({}, meta);
+    if (!connectedIdentityMetadata) {
+      throw new Error(`CryptKeeper: identity metadata is not found`);
+    }
+
+    return connectedIdentityMetadata;
+  };
+
+  newRequest = async (newRequestType: PendingRequestType, newRequestPayload: unknown): Promise<unknown> => {
+    try {
+      const responsePayload = await this.requestManager.newRequest(newRequestType, newRequestPayload);
+      await this.browserService.closePopup();
+
+      return responsePayload;
+    } catch (error) {
+      throw new Error(`${(error as Error).message}`);
+    }
+  };
+
+  requiredApproval = async ({ urlOrigin }: IZkMetadata): Promise<IConnectionApprovalData> => {
+    const { checkedUrlOrigin, isApproved, canSkipApprove } = this.getConnectionApprovalData({ urlOrigin });
+
+    // TODO: This check should be not just `isApproved` but also `isConnected`;
+    // Because the idea is to force to have a connection via `connectedIdentity()`
+    // which includes checking the approval part, this would be done in another PR
+    if (!isApproved) {
+      throw new Error(`CryptKeeper: ${urlOrigin} is not approved, please do a connect request first.`);
+    }
+    await this.checkLock();
+    return { checkedUrlOrigin, isApproved, canSkipApprove };
+  };
+
+  checkApproval = async ({ urlOrigin }: IZkMetadata): Promise<IConnectionApprovalData> => {
+    const { checkedUrlOrigin, isApproved, canSkipApprove } = this.getConnectionApprovalData({ urlOrigin });
+    await this.checkLock();
+    return { checkedUrlOrigin, isApproved, canSkipApprove };
+  };
+
+  private getConnectionApprovalData = ({ urlOrigin }: IZkMetadata): IConnectionApprovalData => {
+    if (!urlOrigin) {
+      throw new Error("CryptKeeper: Origin is not set");
+    }
+
+    const isApproved = this.approvalService.isApproved(urlOrigin);
+    const canSkipApprove = this.approvalService.canSkipApprove(urlOrigin);
+
+    return { checkedUrlOrigin: urlOrigin, isApproved, canSkipApprove };
+  };
+
+  private checkLock = async () => {
```

Comment:
This method name doesn't describe what happens inside. `checkLock` checks if something is locked